### ISSUE TITLE
SSHConnection leaves pending request deferreds on close

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4483]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alanfranzoni reported|
|-|-|
|Trac ID|trac#4483|
|Type|defect|
|Created|2010-06-03 15:50:17Z|
|Branch|https://github.com/twisted/twisted/tree/errback-ssh-deferreds-4483-2|

SSHConnection happens to create deferreds when a sendRequest() or sendGlobalRequest() is called, whenever asking for a reply.

If the connection is closed when waiting for such reply, deferreds are just left hanging and never errbacked nor callbacked.

Attachments:

* [4483.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/806/806a08ec75b433ebda385d624d6878ddb7abc308/bde2a79984075eaee31ddd87b989ac4a5f9e1f62.patch) (7160 bytes) - added by alanfranzoni on 2010-06-03 15:53:41Z - patch for 4483
* [4483_fixreview.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/806/806a08ec75b433ebda385d624d6878ddb7abc308/09d28df96ced2aa5dd6c917f9fb793796a4dd1ef.patch) (4649 bytes) - added by alanfranzoni on 2010-06-04 08:41:15Z - patch after review
* [twisted_errback-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/806/806a08ec75b433ebda385d624d6878ddb7abc308/34fec1910ec193314ea56be58adbc6a9fcfcc9d0.patch) (5277 bytes) - added by alanfranzoni on 2010-06-21 13:59:22Z - Fix false channel testcase

<details><summary>Searchable metadata</summary>

```
trac-id__4483 4483
type__defect defect
reporter__alanfranzoni alanfranzoni
priority__normal normal
milestone__ 
branch__branches_errback_ssh_deferreds_4483_2 branches/errback-ssh-deferreds-4483-2
branch_author__alanfranzoni__exarkun alanfranzoni, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1275580217000000 1275580217000000
changetime__1277763805000000 1277763805000000
version__None None
owner__ 
cc__alanfranzoni
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|alanfranzoni removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [29173]) Branching to 'errback-ssh-deferreds-4483'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [29174]) Apply 4483.patch

refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/errback-ssh-deferreds-4483)

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|alanfranzoni commented|
|-|-|

Mmh, such failures seem to happen on the trunk as well, If I am reading the buildbot results properly, hence I think they're unrelated to this patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to alanfranzoni|
|-|-|

> Mmh, such failures seem to happen on the trunk as well, If I am reading the buildbot results properly, hence I think they're unrelated to this patch.

Yes, that seems right.  The build results look good.

Some other points:

  1. In `twisted/conch/test/test_connection.py`:
    1. `test_channelClosedCausesLeftoverChannelDeferredsToErrback` needs a docstring explaining what it's testing.  This should just describe the behavior being tested.  It doesn't need to say "Test that ..." or anything like that.
    1. You might like `TestCase.assertFailure`, it should simplify some of the assertions being added.
    1. `TestCleanConnectionShutdown` needs a class docstring explaining what it's testing.
    1. The skip for `TestCleanConnectionShutdown` is imprecise.  Is it PyCrypto or pyasn1?  It's unfortunate that the rest of the conch tests are sloppy in this regard.  It would be nice to avoid expanding the problem, though.
    1. The `test_serviceStoppedCausesLeftoverGlobalDeferredsToErrback` docstring doesn't need to say "Tests that ...".  Just say what behavior is being checked.
  1. In `twisted/conch/ssh/connection.py`:
    1. Please add some whitespace to the dict literal in `__init__` - `{foo: [bar]}`
    1. Methods should be separated from each other by two blank lines.  You don't need to fix the whole file, but at least fix it for the new and modified methods.
  1. In `twisted/conch/test/test_ssh.py`:
    1. `unittest.fail` is deprecated.  As long as this code is changing, it'd be nice to switch to a non-deprecated API.  It'd be nice to be able to use `TestCase.assertFailure` here, but of course there's no `TestCase` instance in sight.  It might be sufficient to change these to just `log.err` something.  A logged error will still cause the test to fail.

Thanks again for the patch.  If you can make the next one against the branch, that'd be great.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|alanfranzoni commented|
|-|-|

OK, I'm fixing those during the weekend.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|alanfranzoni removed owner|
|-|-|

Here you are, the patch applies to the branch.

About point 1.4: since twisted X.XX (I don't remember, it's a version after 2.5 that requires pyasn1, beyound pycrypto) the test transport is unavailable unless *both* libraries are available. I changed both docstrings in test connection in order to be more precise.

I hope this is fine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [29188]) Apply 4483_fixreview.patch

refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29189]) Fix some minor coding standard issues

  * use `assertFailure` instead of `failUnlessFailure`
  * Use a `log.err` which will actually fail a test
  * add some docstrings and whitespace


refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to alanfranzoni|
|-|-|

Thanks.  I made a couple more small changes, and I think this mostly looks good now.  One more thing, though.  Somehow I was convinced the `test_ssh.py` change was correct before, but now I'm not sure.  Can you explain the motivation behind that a bit?

It seems to me that the exec false _should_ still get a reply.  The fact that the tests pass now with the assertion inverted seems to suggest to me that something is going wrong.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|alanfranzoni set owner to @exarkun|
|-|-|

The real answer is: I 'm not sure. I don't understand what are the assertions of such test case. It seems a kind of integration test of some sort, without clear expectations beyond "try doing that and verify no errors are raised"

Test this patch on the trunk:

```
diff --git a/twisted/conch/test/test_ssh.py b/twisted/conch/test/test_ssh.py
index c165870..016acaa 100644
--- a/twisted/conch/test/test_ssh.py
+++ b/twisted/conch/test/test_ssh.py
@@ -457,7 +457,7 @@ if Crypto is not None and pyasn1 is not None:
             log.msg('opened false')
 
         def _cbRequestWorked(self, ignored):
-            pass
+            unittest.fail('false exec succeeded')
 
         def _ebRequestFailed(self, reason):
             unittest.fail('false exec failed: %s' % reason)
```

Setting a unittest.fail on the callback as well as the errback will NOT make the test fail, because such deferred won't (on the current trunk) errback nor callback, i.e. it is left hanging - which is the very problem I'm trying to address with my patch. As soon as the underlying problem is addressed, such test expectations must be corrected.

Since FalseTransport seems to drop the connection as soon as it's established, I think it is correct to expect a failure from the request on it, but again - I'm not sure because I don't know what's under test here.

Side note:
 * TestCase.assertFailure seems an alias of TestCase.failUnlessFailure, and it's unlisted on the API page:

 http://twistedmatrix.com/documents/current/api/twisted.trial.unittest.TestCase.html

 Why should it be preferred over non-aliased, documented methods?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

I investigated a bit further, and it looks like there's a bug in the code covered by the modified `test_ssh.py` methods.  I'm going to try to fix that first.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|alanfranzoni commented|
|-|-|

Drop a comment should you need an hand.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

> Drop a comment should you need an hand.

Thanks.  I filed [#4490](https://github.com/twisted/twisted/issues/4490), going to wait on that before trying to finish this.  I think it'll reveal another bug in test_ssh which is exposed by firing these Deferreds, but I'm not sure yet.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|alanfranzoni removed owner|
|-|-|

[#4490](https://github.com/twisted/twisted/issues/4490) has now been closed and I've tried merging changes from trunk and it seems that, beyond a conflict on twisted/conch/ssh/connection.py that can be solved by just choosing mf (mine-full) as conflict resolution method, it still works fine.

Is there any work left on this branch before it can be merged back?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [29320]) Branching to 'errback-ssh-deferreds-4483-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to alanfranzoni|
|-|-|

After working on [#4490](https://github.com/twisted/twisted/issues/4490), I'm now pretty sure that the `SSHTestFalseChannel` change isn't correct.  Opening a channel and exec'ing `false` on it should succeed.  So the reason for the incorrect errback from the `sendRequest` in `SSHTestFlaseChannel.channelOpen` needs to be tracked down and corrected.  Or someone needs to explain why I'm mistaken.  (I might try to do that tracking down, but it might take a little while before I get to it.)

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|alanfranzoni set owner to @exarkun|
|-|-|

This was pretty hard to track down. It looked like a timing issue, since just adding a write() call in the FalseTransport made the problem disappear.

You were right about the test; the request is supposed to succeed. I tested that very case using an OpenSSH server and it resulted in a request success, but it didn't raise the "Channel closed" exception either! 

That made me think the problem lied in the test case rather than in the code.

What happens is:

if loseConnection() is called on the transport while there's something to be written (or being written) to the other side, MSG_CHANNEL_CLOSE is not sent immediately; instead, a flag (self.closing) is set, everything is written, then MSG_CHANNEL_CLOSE gets sent. This is what happens with all other test cases, all of which make use of write()

On the contrary, FalseTransport writes *nothing* to the other side; this causes MSG_CHANNEL_CLOSE to be sent immediately to the other side, causing the channel to be closed and the "Channel closed" exception to be raised.

In fact, on the server side when using FalseTransport the control flow for the exec request is:
 * twisted.conch.ssh.connection.SSHConnection.ssh_CHANNEL_REQUEST is invoked when the exec request from the client arrives, and it calls channel.requestReceived(), then adding a callback and an errback to its result.
 * this causes twisted.conch.ssh.session.SSHSession.request_exec to be invoked;
* such method causes test_ssh.ConchSessionForTestAvatar.execCommand() to be invoked;
* inside that method, FalseTransport is instantiated  and it's immediately asked to inovoke SSHSessionProcessProtocol.processEnded with an exit status of 255
* in turn processEnded() calls loseConnection() on its channel, and a MSG_CHANNEL_CLOSE is immediately sent on the other side. As soon as the client receives such message, having a pending request in its self.deferreds attribute, raises a "Channel closed" ConchError as per my patch.
* at that point we get back to twisted.conch.ssh.connection.ssh_CHANNEL_REQUEST callback, _cbChannelRequest, which tries to send a MSG_CHANNEL_SUCCESS, but the channel is already closed at that point and nothing can change its state.


My twisted_errback-2.patch offers an inelegant workaround for such issue.

I think the "real" issue here is that test_ssh integration tests are both clumsy and coupled. I'd be happy to contribute a better version of such tests, but that's pretty outside the scope of this ticket. I hope the patch can be applied nonetheless.

**Comment by twisted-trac** [comment]
|<img alt="alanfranzoni's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|alanfranzoni commented|
|-|-|

SIgh, it seems the wiki formatting got broken at some point, I'm sorry but I seem to have no comment editing permission.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_23"></a>|@thijstriemstra commented|
|-|-|

Replying to [alanfranzoni](#note_22):
> SIgh, it seems the wiki formatting got broken at some point, I'm sorry but I seem to have no comment editing permission.

Unfortunately Trac doesn't allow anyone to modify comments from the web interface, so no worries :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [29368]) Apply twisted_errback-2.patch

refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for the further investigation!  The new changes look safer, although it's too bad Conch doesn't deal with this re-entrancy (and too bad the test suite was doing it anyway, and too bad no one noticed for however many years it's been there).

There are a few cosmetic changes left to make, but I'll take care of those and then merge the branch.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [29390]) Merge errback-ssh-deferreds-4483-2

Author: alanfranzoni
Reviewer: exarkun
Fixes: [#4483](https://github.com/twisted/twisted/issues/4483)


Fix `twisted.conch.ssh.connection.SSHConnection` so that when it is disconnected, it fires any
remaining `Deferred`s previously returned from a `sendRequest` or `sendGlobalRequest` call which
had not yet been fired.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [29400]) Branching to 'faster-close-fds-4483'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [29401]) Apply 0002-Changes-from-James-review.patch

refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

Sorry, wrong ticket number.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

(In [29404]) Apply 0002-Changes-from-James-review.patch

refs [#4483](https://github.com/twisted/twisted/issues/4483)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
